design/cors-configuration.md

```diff
@@ -0,0 +1,179 @@
+- Requests do not set custom headers, such as X-Other-Header.
+
+If the content of the request doesn't meet the criteria above, the browser first checks whether the actual request should be sent. This is done by sending a special request (called preflight request) to the server in advance. A preflight request first sends an HTTP request to the resource using the OPTIONS method with the following headers:
+- `Origin`: Specifies the domain that would like access to the resource. This is inserted by the browser in a cross-origin request.
+- `Access-Control-Request-Method`: The HTTP method to be used in the actual request from the browser.
+- `Access-Control-Request-Headers`: The custom headers to be sent in the actual cross-origin request.
+
+In response to a preflight request, the server sends a response with the following headers:
+- `Access-Control-Allow-Origin`: Specifies the domains allowed to access the resource.
+- `Access-Control-Allow-Credentials`: Indicates whether browser credentials can be used to make the actual request (cookies for instance).
+- `Access-Control-Expose-Headers`: Allows headers to be exposed to the browser.
+- `Access-Control-Max-Age`: Specifies how long preflight request results can be cached.
+- `Access-Control-Allow-Methods`: Indicates which methods are allowed when making an actual request.
+- `Access-Control-Allow-Headers`: Indicates which headers can be used in the actual request.
+
+When the browser gets the preflight response, it checks if the origin is allowed and if the HTTP method and headers of the main request are in the list returned by the server. If so, it sends the main request, which will be a regular cross-origin request, it will include the `Origin` header and the response will contain `Access-Control-Allow-Origin` once again.
+
+This proposal introduces a way to set all the CORS related configuration in Contour, letting Envoy do all the heavy work.
+
+## High-Level Design
+
+Envoy supports [CORS via a filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/cors_filter.html) and it can be configured [using the API](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-corspolicy). The changes proposed in this document will allow the configuration of CORS policies in Contour.
+
+At a high level the proposed changes will imply:
+- Adding new fields at virtual host level to configure the CORS policy in the YAML.
+- Changing some structs in the code.
+- Enabling the Envoy CORS filter.
+
+### Proposed YAML fields
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+ name: google
+ namespace: prod
+spec:
+ virtualhost:
+   fqdn: www.google.com
+   # The CORS policy described here is local to match and apply for "/finance" only.
```

Comment:
With the way that HTTPProxy works, this implies that all Routes under this virtualhost will have the same CORS policy.

I'm not familiar enough with how CORS is used for real to see if that's bad. I suspect it may lead to unintended consequences maybe?

Sorry to bikeshed more on this one.

Reply:
I updated the example, hopefully it's going to be easyer now to understand the way CORS policy will be applied.

The key here is that usually the APIs are exposed in their own domain. If we take a look at the way [Google publishes its APIs](https://developers.google.com/apis-explorer) we see that they use the domain https://www.googleapis.com:

- https://www.googleapis.com/analytics/ for analytics related APIs
- https://www.googleapis.com/blogs/ for blogger related APIs
...

All the APIs at https://www.googleapis.com need the same CORS Policy, so having the ability to set the policy at virtual host level comes in handy. IMHO it would be weird that someone needed to set a different policy for a certain route under the same domain. If that happened, a workaround could be to use a subdomain for that API, something like https://internal.googleapis.com (this is not a real API). For some reason, Google uses subdomains for some of their APIs (see https://bigquery.googleapis.com for instance) but I don't think they do this because of CORS.

For that reason, I think that is pretty safe to start with the HTTPProxy approach and see how the policy is being used. If we see that disabling the policy at route level is necessary we could add it afterwards. 

Just in case, @yoitsro, @glerchundi, do you know any real world example where different paths need different CORS policies for the same domain?

Reply:
In our experience, CORS at the virtual host level is completely fine. As you say, if someone does need to have a different policy at a route level, that'd be out of the ordinary.